File: length_of_last_word.py
"""
58. Length of Last Word

Time Complexity: O(n)
Space Complexity: O(1)
"""


class Solution:
    def lengthOfLastWord(self, s: str) -> int:
        i = len(s) - 1
        length = 0
        while i >= 0 and s[i] == ' ':
            i -= 1
        while i >= 0 and s[i] != ' ':
            length += 1
            i -= 1
        return length


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    s1 = "Hello World"
    print(f"Example 1: {solution.lengthOfLastWord(s1)}")  # Expected: 5
    
    # Example 2
    s2 = "   fly me   to   the moon  "
    print(f"Example 2: {solution.lengthOfLastWord(s2)}")  # Expected: 4
    
    # Example 3
    s3 = "luffy is still joyboy"
    print(f"Example 3: {solution.lengthOfLastWord(s3)}")  # Expected: 6
    
    # Edge case: single word
    s4 = "Hello"
    print(f"Example 4: {solution.lengthOfLastWord(s4)}")  # Expected: 5
    
    # Edge case: single word with trailing spaces
    s5 = "a   "
    print(f"Example 5: {solution.lengthOfLastWord(s5)}")  # Expected: 1
    
    # Edge case: single word with leading spaces
    s6 = "   a"
    print(f"Example 6: {solution.lengthOfLastWord(s6)}")  # Expected: 1
    
    # Edge case: multiple spaces between words
    s7 = "a  b    c"
    print(f"Example 7: {solution.lengthOfLastWord(s7)}")  # Expected: 1
    
    # Edge case: long last word
    s8 = "a programming"
    print(f"Example 8: {solution.lengthOfLastWord(s8)}")  # Expected: 11
    
    # Test alternative method
    s9 = "   fly me   to   the moon  "
    print(f"Example 9 (alternative): {solution.lengthOfLastWordAlternative(s9)}")  # Expected: 4

File: length_of_last_word.cpp
/*
58. Length of Last Word

Time Complexity: O(n)
Space Complexity: O(1)
*/

#include <iostream>
#include <string>
#include <sstream>
#include <vector>
using namespace std;


class Solution {
public:
    int lengthOfLastWord(string s) {
        int i = s.length() - 1;
        int length = 0;
        while (i >= 0 && s[i] == ' ') {
            i--;
        }
        while (i >= 0 && s[i] != ' ') {
            length++;
            i--;
        }
        return length;
    }
};


// Test cases
int main() {
    Solution solution;
    
    // Example 1
    cout << "Example 1: " << solution.lengthOfLastWord("Hello World") << endl;  // Expected: 5
    
    // Example 2
    cout << "Example 2: " << solution.lengthOfLastWord("   fly me   to   the moon  ") << endl;  // Expected: 4
    
    // Example 3
    cout << "Example 3: " << solution.lengthOfLastWord("luffy is still joyboy") << endl;  // Expected: 6
    
    // Edge case: single word
    cout << "Example 4: " << solution.lengthOfLastWord("Hello") << endl;  // Expected: 5
    
    // Edge case: single word with trailing spaces
    cout << "Example 5: " << solution.lengthOfLastWord("a   ") << endl;  // Expected: 1
    
    // Edge case: single word with leading spaces
    cout << "Example 6: " << solution.lengthOfLastWord("   a") << endl;  // Expected: 1
    
    // Edge case: multiple spaces between words
    cout << "Example 7: " << solution.lengthOfLastWord("a  b    c") << endl;  // Expected: 1
    
    // Edge case: long last word
    cout << "Example 8: " << solution.lengthOfLastWord("a programming") << endl;  // Expected: 11
   
    return 0;
}
